### PR TITLE
Builder implemented for MinioClient, Arguments

### PR DESCRIPTION
**Pull request by BigUstad**
Request & Response processing.
Versioning added as an example of using MinioClient/Args builder.

### CODE REVIEW

File: Minio/DataModel/Args.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+using System.Collections;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using RestSharp;
+
+namespace Minio
+{
+    public class Args
+    {
+        internal Hashtable ExtraHeaders { get; set; }
```
Comment (poornas) [inline Minio/DataModel/Args.cs:26]: ExtraHeaders -> Headers

File: Minio/DataModel/Args.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        internal Hashtable ExtraQueryParams { get; set; }
```
Comment (poornas) [inline Minio/DataModel/Args.cs:27]: ExtraQueryParams -> QueryParams

File: Minio/DataModel/BucketOperationsArgs.cs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        public new BucketExistsArgs WithLocation(string loc)
+        {
+            this.Location = loc;
+            return this;
+        }
+        public BucketExistsArgs WithRegion(string reg)
+        {
+            this.Region = reg;
+            return this;
+        }
```
Comment (poornas) [inline Minio/DataModel/BucketOperationsArgs.cs:44]: All the above methods excepting GetRequest will be inherited from BucketArgs, can be removed here.

File: Minio/DataModel/BucketOperationsArgs.cs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        public new RemoveBucketArgs WithLocation(string loc)
+        {
+            this.Location = loc;
+            return this;
+        }
+        public RemoveBucketArgs WithRegion(string reg)
+        {
+            this.Region = reg;
+            return this;
+        }
```
Comment (poornas) [inline Minio/DataModel/BucketOperationsArgs.cs:72]: All the above methods will be inherited from BucketArgs, no need to repeat.

File: Minio/DataModel/BucketOperationsArgs.cs [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        public new ListObjectsArgs WithLocation(string loc)
+        {
+            this.Location = loc;
+            return this;
+        }
+        public ListObjectsArgs WithRegion(string reg)
+        {
+            this.Region = reg;
+            return this;
+        }
```
Comment (poornas) [inline Minio/DataModel/BucketOperationsArgs.cs:95]: All the above methods will be inherited from BucketArgs, can be removed.

File: Minio/DataModel/BucketOperationsArgs.cs [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        public new MakeBucketArgs WithLocation(string loc)
+        {
+            this.Location = loc;
+            return this;
+        }
+        public MakeBucketArgs WithRegion(string reg)
+        {
+            this.Region = reg;
+            return this;
+        }
```
Comment (poornas) [inline Minio/DataModel/BucketOperationsArgs.cs:130]: remove above methods as they are inherited.

File: Minio/DataModel/BuilderUtil.cs [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+            }
+
+            // Check host if it is a hostname.
+            if (Uri.CheckHostName(host).ToString().ToLower().Equals("dns"))
+            {
+                return true;
+            }
+            return false;
+        }
+    }
```
Comment (poornas) [inline Minio/DataModel/BuilderUtil.cs:136]: these methods need unit tests.

File: Minio/DataModel/VersioiningConfiguration.cs [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System;
+using System.Xml.Serialization;
```
Comment (poornas) [inline Minio/DataModel/VersioiningConfiguration.cs:18]: file name change from Minio/DataModel/VersioiningConfiguration.cs -> Minio/DataModel/VersioningConfiguration.cs

File: Minio/DataModel/VersioningArgs.cs [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        {
+        }
+
+
+        public new void Validate()
+        {
+            utils.ValidateBucketName(this.BucketName);
+            if ( this.VersioningSuspended && !this.VersioningEnabled )
+            {
+                throw new InvalidBucketOperationException(this.BucketName, "Versioning has to be enabled to be suspended.");
```
Comment (poornas) [inline Minio/DataModel/VersioningArgs.cs:84]: this doesn't make sense. VersioningEnabled and VersioningSuspended are toggled states - you don't need 2 variables to represent that.

**Comment by poornas** [comment]
package references seems broken - see a lot of compilation errors.
```
MinioClient.cs(365,107): error CS0518: Predefined type 'System.String' is not defined or imported [/home/kp/code/minio-dotnet/Minio/Minio.csproj]
MinioClient.cs(365,133): error CS0518: Predefined type 'System.String' is not defined or imported [/home/kp/code/minio-dotnet/Minio/Minio.csproj]
MinioClient.cs(440,184): error CS0518: Predefined type 'System.Object' is not defined or imported [/home/kp/code/minio-dotnet/Minio/Minio.csproj]
MinioClient.cs(440,184): error CS0246: The type or namespace name 'CancellationToken' could not be found (are you missing a using directive or an assembly reference?) [/home/kp/code/minio-dotnet/Minio/Minio.csproj]
```

File: Minio/DataModel/MinioClientBuilder.cs [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+using System;
+using System.Net;
+using System.Net.Http;
+using System.Security.Authentication;
+using System.Security.Cryptography.X509Certificates;
+using Minio.Exceptions;
+
+namespace Minio
+{
+    public interface IBuilder
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/DataModel/MinioClientBuilder.cs:27]: Changing.

File: Minio/DataModel/MinioClientBuilder.cs [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+                InitClientBuilder();
+            }
+            return this;
+        }
+
+        public MinioClient WithEndpoint(Uri url)
+        {
+            if (url != null )
+            {
+                throw new ArgumentException(String.Format("URL is null. Can't create endpoint."));
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/DataModel/MinioClientBuilder.cs:121]: Input sanity check for use in the call.

File: Minio/DataModel/BucketOperationsArgs.cs (hunk 4 above)
Comment (BigUstad) [reply Minio/DataModel/BucketOperationsArgs.cs:72]: It is to get the derived return type to be used with `new BucketExistsArgs()`. I've made the change accordingly to use the base method.

File: Minio/DataModel/BucketOperationsArgs.cs (hunk 5 above)
Comment (BigUstad) [reply Minio/DataModel/BucketOperationsArgs.cs:95]: Changed to call base method with `new`.

File: Minio/DataModel/BucketOperationsArgs.cs (hunk 6 above)
Comment (BigUstad) [reply Minio/DataModel/BucketOperationsArgs.cs:130]: Changed to call base method with `new`.

File: Minio/DataModel/BuilderUtil.cs (hunk 7 above)
Comment (BigUstad) [reply Minio/DataModel/BuilderUtil.cs:136]: I will add it.

File: Minio/DataModel/VersioiningConfiguration.cs (hunk 8 above)
Comment (BigUstad) [reply Minio/DataModel/VersioiningConfiguration.cs:18]: Changed.

File: Minio/DataModel/VersioningArgs.cs (hunk 9 above)
Comment (BigUstad) [reply Minio/DataModel/VersioningArgs.cs:84]: Implemented based on a previous conversation with @kannappanr . States are Versioning not Enabled (versioning configuration is null, hence both flags false), Versioning Enabled and not suspended (enabled flag is true, suspended is false), Versioning Enabled but suspended (enabled is true, suspended is also true). So, suspended true & enabled false is an illegitimate status. Please do let me know if this implementation is correct.

**Comment by BigUstad** [comment]
> package references seems broken - see a lot of compilation errors.
> 
> ```
> MinioClient.cs(365,107): error CS0518: Predefined type 'System.String' is not defined or imported [/home/kp/code/minio-dotnet/Minio/Minio.csproj]
> MinioClient.cs(365,133): error CS0518: Predefined type 'System.String' is not defined or imported [/home/kp/code/minio-dotnet/Minio/Minio.csproj]
> MinioClient.cs(440,184): error CS0518: Predefined type 'System.Object' is not defined or imported [/home/kp/code/minio-dotnet/Minio/Minio.csproj]
> MinioClient.cs(440,184): error CS0246: The type or namespace name 'CancellationToken' could not be found (are you missing a using directive or an assembly reference?) [/home/kp/code/minio-dotnet/Minio/Minio.csproj]
> ```

I'm not seeing these errors on my setup. Maybe Nuget installation of System.Net.Http is needed.  These lines have code that I've not changed in this PR

File: Minio.Examples/Program.cs [hunk 12]
```diff
@@ -84,13 +84,16 @@ public static void Main(string[] args)
...
+            MinioClient minioClient_ver = null;
             if (enableHTTPS)
             {
                 minioClient = new MinioClient(endPoint, accessKey, secretKey).WithSSL();
+                minioClient_ver = new MinioClient().WithEndpoint(endPoint).WithCredentials(accessKey, secretKey).WithSSL().Build();
             }
             else
             {
                 minioClient = new MinioClient(endPoint, accessKey, secretKey);
+                minioClient_ver = new MinioClient().WithEndpoint(endPoint).WithCredentials(accessKey, secretKey).Build();
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio.Examples/Program.cs:96]: Yes. Now changed.

File: Minio/DataModel/BucketOperationsArgs.cs (hunk 3 above)
Comment (BigUstad) [reply Minio/DataModel/BucketOperationsArgs.cs:44]: Changed to call base method with `new`.

File: Minio.Examples/Program.cs [hunk 13]
```diff
@@ -84,13 +91,28 @@ public static void Main(string[] args)
 
             // WithSSL() enables SSL support in MinIO client
             MinioClient minioClient = null;
-            if (enableHTTPS)
+            if ( port > 0 )
```
Comment (poornas) [inline Minio.Examples/Program.cs:94]: there is no need for so many client construction methods. If port is not set and https set 443 otherwise 80 if unspecified.

File: Minio/DataModel/BucketOperationsArgs.cs [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+        {
+            return (ListObjectsArgs)(base.WithLocation(loc));
+        }
+        public new ListObjectsArgs WithRegion(string reg)
+        {
+            return (ListObjectsArgs)(base.WithRegion(reg));
+        }
+        public new RestRequest GetRequest(string baseUrl, RestSharp.Authenticators.IAuthenticator authenticator)
+        {
+            return RequestUtil.CreateRequest(baseUrl, Method.DELETE, authenticator, this.BucketName, this.Secure,  this.Region);
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/DataModel/BucketOperationsArgs.cs:90]: It is a function that returns a RestRequest object. RemoveBucket will return a Method.DELETE request. Shall I change the function name if there is confusion?

File: Minio.Examples/Program.cs (hunk 13 above)
Comment (BigUstad) [reply Minio.Examples/Program.cs:94]: Implemented this change.

File: Minio/ApiEndpoints/BucketOperations.cs [hunk 15]
```diff
@@ -56,6 +56,104 @@ public async Task<ListAllMyBucketsResult> ListBucketsAsync(CancellationToken can
             return bucketList;
         }
 
+        /// <summary>
+        /// Check if a private bucket with the given name exists.
+        /// </summary>
+        /// <param name="args">Arguments Object that has the info</param>
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/ApiEndpoints/BucketOperations.cs:62]: Yes. Changed.

File: Minio/ApiEndpoints/BucketOperations.cs [hunk 16]
```diff
@@ -56,6 +56,104 @@ public async Task<ListAllMyBucketsResult> ListBucketsAsync(CancellationToken can
...
+
+        /// <param name="args">Arguments Object that has the info</param>
+        /// <param name="cancellationToken">Optional cancellation token to cancel the operation</param>
+        /// <returns> Task </returns>
+        /// <exception cref="InvalidBucketNameException">When bucketName is invalid</exception>
+        /// <exception cref="InvalidBucketOperationException">When Args object or another setting is wrong</exception>
+        public async Task  SetVersioningAsync(SetVersioningArgs args, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            Uri requestUrl = RequestUtil.MakeTargetURL(this.BaseUrl, this.Secure, args.Location);
+            this.SetTargetURL(requestUrl);
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/ApiEndpoints/BucketOperations.cs:150]: Not all of these member variables are set in Args object/ child classes. Hence I've used it here.

File: Minio/DataModel/BucketOperationsArgs.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (poornas) [reply Minio/DataModel/BucketOperationsArgs.cs:90]: ListObjectsArgs should use Method.GET here

File: Minio/DataModel/VersioningArgs.cs (hunk 9 above)
Comment (poornas) [reply Minio/DataModel/VersioningArgs.cs:79]: can you rename this.VersioningEnabled -> this.Versioned to make it clearer?

File: Minio.Examples/Program.cs [hunk 17]
```diff
@@ -77,22 +88,18 @@ public static void Main(string[] args)
                 accessKey = "Q3AM3UQ867SPQQA43P2F";
                 secretKey = "zuf+tfteSlswRu7BJ86wekitnifILbZam1KYY3TG";
                 enableHTTPS = true;
+                port = 443;
```
Comment (kannappanr) [inline Minio.Examples/Program.cs:91]: why do we need `port` for standard ports like 443 and 80?

File: Minio/ApiEndpoints/ObjectOperations.cs [hunk 18]
```diff
@@ -272,7 +272,7 @@ public async Task PutObjectAsync(string bucketName, string objectName, Stream da
 
                 }
             }
-            
+
```
Comment (kannappanr) [inline Minio/ApiEndpoints/ObjectOperations.cs:275]:
```suggestion
```

File: Minio/Helper/RequestUtil.cs [hunk 19]
```diff
@@ -145,5 +147,82 @@ internal static bool IsValidEndpoint(string endpoint)
...
+            // Start with user specified endpoint
+            string host = baseURL;
+
+            string resource = string.Empty;
+            bool usePathStyle = false;
+            if (bucketName != null)
+            {
+                if (s3utils.IsAmazonEndPoint(baseURL))
+                {
+                    usePathStyle = false;
```
Comment (kannappanr) [inline Minio/Helper/RequestUtil.cs:172]:
```suggestion
```

File: Minio/DataModel/VersioningArgs.cs (hunk 9 above)
Comment (BigUstad) [reply Minio/DataModel/VersioningArgs.cs:79]: Change done.

File: Minio.Examples/Program.cs (hunk 17 above)
Comment (BigUstad) [reply Minio.Examples/Program.cs:91]: I made the change for more brevity in code. [Suggestion](https://github.com/minio/minio-dotnet/pull/446#discussion_r466103780).

File: Minio/ApiEndpoints/ObjectOperations.cs (hunk 18 above)
Comment (BigUstad) [reply Minio/ApiEndpoints/ObjectOperations.cs:275]: Implemented in all the files (in this patchset only) in which this was present.

File: Minio/DataModel/BucketOperationsArgs.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (poornas) [reply Minio/DataModel/BucketOperationsArgs.cs:103]: line 103 should have a Method.GET - ListObjects request url would not be sending a Delete request.

File: Minio/MinioClient.cs [hunk 20]
```diff
@@ -376,6 +435,14 @@ internal void SetTargetURL(Uri uri)
             this.restClient.BaseUrl = uri;
         }
 
+        /// <summary>
+        /// Get If we are using a secure connection
+        /// </summary>
+        public bool GetSecured()
```
Comment (poornas) [inline Minio/MinioClient.cs:441]:
```suggestion
        public bool IsSecure()
```

Comment (BigUstad) [reply Minio/MinioClient.cs:441]: Changed.

File: Minio/DataModel/BucketArgs.cs [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+        internal bool Versioned { get; set; }
+        internal bool VersioningEnabled { get; set; }
+        internal bool VersioningSuspended { get; set; }
+
+        public new void Validate()
+        {
+            utils.ValidateBucketName(this.BucketName);
+            if ( !this.Versioned && (this.VersioningEnabled || this.VersioningSuspended) )
+            {
+                throw new UnexpectedMinioException(String.Format("{0} or {1} cannot be enabled without {2} being enabled.", VersioningEnabled, VersioningSuspended, Versioned ));
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/DataModel/BucketArgs.cs]: Changed the message. Also validation now in VersionArgs, unlikely it will be set in BucketArgs.

**Comment by poornas** [comment]
Docs need to be updated for the versioning API's, other than that LGTM.

**Comment by BigUstad** [comment]
> Docs need to be updated for the versioning API's, other than that LGTM.

I pushed the docs updated.

File: Minio/DataModel/BucketOperationsArgs.cs [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+        }
+        public new RestRequest GetRequest(string baseUrl, RestSharp.Authenticators.IAuthenticator authenticator)
+        {
+            return RequestUtil.CreateRequest(baseUrl, Method.GET, authenticator, this.BucketName, this.Secure,  this.Region);
+        }
+        public new ListObjectsArgs WithSSL(bool secure=true)
+        {
+            return (ListObjectsArgs)(base.WithSSL(secure));
+        }
+    }
```
Comment (poornas) [inline Minio/DataModel/BucketOperationsArgs.cs:109]: for bucket operations, only MakeBucket requires location and region. Can you remove the WithLocation and WithRegion methods from all the remaining operations.

Comment (poornas) [reply Minio/DataModel/BucketOperationsArgs.cs:109]: docs is missing example for ListObjectsAsync with this args variant.

Comment (poornas) [reply Minio/DataModel/BucketOperationsArgs.cs:109]: can you also change the functional tests and example programs to this new style.

Comment (BigUstad) [reply Minio/DataModel/BucketOperationsArgs.cs:109]: ListObjectsAsync isn't implemented yet. Everything else it needs is implemented. I'll implement ListObjectsAsync in the next PR.

File: Minio.Examples/Cases/RemoveBucket.cs [hunk 23]
```diff
@@ -23,11 +23,17 @@ class RemoveBucket
...
+                                     string region="us-east-1")
         {
             try
             {
-                await minio.RemoveBucketAsync(bucketName);
+                await minio.RemoveBucketAsync(
+                    new RemoveBucketArgs()
+                            .WithBucket(bucketName)
+                            .WithRegion(region)
+                            .WithSSL(minio.IsSecure())
```
Comment (kannappanr) [inline Minio.Examples/Cases/RemoveBucket.cs:35]: Why is this here?

File: Minio/DataModel/BucketOperationsArgs.cs (hunk 22 above)
Comment (BigUstad) [reply Minio/DataModel/BucketOperationsArgs.cs:109]: Some cases (MakeBucket, RemoveBucket, BucketExists and Versioning) are implemented. Also Unit Tests.  Changing functional test(Program.cs).

File: Minio/ApiEndpoints/BucketOperations.cs [hunk 24]
```diff
@@ -56,6 +56,100 @@ public async Task<ListAllMyBucketsResult> ListBucketsAsync(CancellationToken can
...
+        /// <param name="cancellationToken">Optional cancellation token to cancel the operation</param>
+        /// <returns> Task </returns>
+        /// <exception cref="InvalidBucketOperationException">When Args object or another setting is wrong</exception>
+        public async Task RemoveBucketAsync(RemoveBucketArgs args, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            args.Validate();
+            this.restClient.Authenticator = new V4Authenticator(this.Secure, this.AccessKey, this.SecretKey, region: this.Region, sessionToken: this.SessionToken);
+            var request = args.GetRequest(this.BaseUrl, this.restClient.Authenticator);
+            var bktRegion = await GetRegion(args.BucketName).ConfigureAwait(false);
+            Uri requestUrl = RequestUtil.MakeTargetURL(this.BaseUrl, this.Secure, args.BucketName, region: bktRegion , usePathStyle: false);
```
Comment (poornas) [inline Minio/ApiEndpoints/BucketOperations.cs:101]: this entire code block - lines 98 thru 101 should all be factored into CreateRequest. It is repeated across API calls. CreateRequest can take Args as a parameter, and cast to specific type for GetRequest

Comment (BigUstad) [reply Minio/ApiEndpoints/BucketOperations.cs:101]: Yes, I'll do that change.

File: Docs/API.md [hunk 25]
```diff
@@ -105,6 +188,12 @@ MinioClient s3Client = new MinioClient("s3.amazonaws.com").WithSSL();
 MinioClient s3Client = new MinioClient("s3.amazonaws.com:80",
                                        accessKey:"YOUR-ACCESSKEYID",
                                        secretKey:"YOUR-SECRETACCESSKEY").WithSSL();
+// 3. Using Builder with public MinioClient(), Endpoint, Credentials & Secure connection
+MinioClient minioClient = new MinioClient()
+                                    .WithEndpoint("s3.amazonaws.com:80")
```
Comment (kannappanr) [inline Docs/API.md:193]:
```suggestion
                                    .WithEndpoint("s3.amazonaws.com")
```

File: Docs/API.md [hunk 26]
```diff
@@ -105,6 +188,12 @@ MinioClient s3Client = new MinioClient("s3.amazonaws.com").WithSSL();
 MinioClient s3Client = new MinioClient("s3.amazonaws.com:80",
```
Comment (kannappanr) [inline Docs/API.md:188]:
```suggestion
MinioClient s3Client = new MinioClient("s3.amazonaws.com",
```

File: Minio/ApiEndpoints/ObjectOperations.cs [hunk 27]
```diff
@@ -1240,4 +1240,4 @@ public async Task<string> PresignedPutObjectAsync(string bucketName, string obje
             return Tuple.Create(this.restClient.BaseUrl.AbsoluteUri, policy.GetFormData());
         }
     }
-}
+}
```
Comment (kannappanr) [inline Minio/ApiEndpoints/ObjectOperations.cs:1243]: Please remove changes in this file

Comment (BigUstad) [reply Minio/ApiEndpoints/ObjectOperations.cs:1243]: In a previous comment you asked to remove newlines at the end of the file.

Comment (BigUstad) [reply Minio/ApiEndpoints/ObjectOperations.cs:1243]: Will do.

File: Minio/DataModel/MinioClientBuilder.cs [hunk 28]
```diff
@@ -0,0 +1,195 @@
...
+            {
+                url = new Uri(conn_url);
+                string hostnameOfUri = url.Authority;
+
+                if (!BuilderUtil.IsValidHostnameOrIPAddress(hostnameOfUri))
+                {
+                    throw new InvalidEndpointException(String.Format("{0}, {1} is invalid hostname.", endpoint, hostnameOfUri),"endpoint");
+                }
+            }
+            catch (Exception ex)
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/DataModel/MinioClientBuilder.cs:100]: Yes. Changed.

File: Minio/ApiEndpoints/BucketOperations.cs [hunk 29]
```diff
@@ -56,6 +56,88 @@ public async Task<ListAllMyBucketsResult> ListBucketsAsync(CancellationToken can
...
+        /// </summary>
+        /// <param name="args">MakeBucketArgs Arguments Object that has bucket info like name, region, location. etc</param>
+        /// <param name="cancellationToken">Optional cancellation token to cancel the operation</param>
+        /// <returns> Task </returns>
+        /// <exception cref="InvalidBucketNameException">When bucketName is invalid</exception>
+        /// <exception cref="InvalidBucketOperationException">When Args object or another setting is wrong</exception>
+        public async Task MakeBucketAsync(MakeBucketArgs args, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            args.Validate();
+            this.UpdateArgs(args);
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/ApiEndpoints/BucketOperations.cs:109]: OK. Changed.

File: Minio/DataModel/MinioClientBuilder.cs [hunk 30]
```diff
@@ -0,0 +1,195 @@
...
+            if (string.IsNullOrEmpty(this.AccessKey) || string.IsNullOrEmpty(this.SecretKey) )
+            {
+                throw new MinioException("User Access Credentials not initialized.");
+            }
+
+            string host = this.BaseUrl;
+
+            var scheme = this.Secure ? utils.UrlEncode("https") : utils.UrlEncode("http");
+
+            if ( !this.BaseUrl.Contains("http") )
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/DataModel/MinioClientBuilder.cs:175]: Changed the call.

File: Minio/DataModel/GenericResponse.cs [hunk 31]
```diff
@@ -0,0 +1,35 @@
...
+
+using RestSharp;
+using System;
+
+namespace Minio
+{
+    public class GenericResponse
+    {
+        internal IRestResponse Response { get; private set; }
+        public GenericResponse Init(IRestResponse res, Args args)
```
Comment (kannappanr) [inline Minio/DataModel/GenericResponse.cs:25]: Why do you need Args?

File: Minio/DataModel/VersioningArgs.cs [hunk 32]
```diff
@@ -0,0 +1,162 @@
...
+            return request;
+        }
+
+        public override RestRequest BuildRequest(RestRequest req)
+        {
+            req.AddQueryParameter("versioning","");
+            return req;
+        }
+
+        public VersioningConfiguration ProcessResponse(RestSharp.IRestResponse response)
```
[the remark this answers is not in the document]
Comment (kannappanr) [reply Minio/DataModel/VersioningArgs.cs:39]: Are we not removing this method?

File: Minio/MinioClient.cs [hunk 33]
```diff
@@ -142,6 +145,42 @@ private async Task<string> GetRegion(string bucketName)
...
         }
 
+        /// <summary>
+        /// Constructs a RestRequest. For AWS, this function has the side-effect of overriding the baseUrl
+        /// in the RestClient with region specific host path or virtual style path.
+        /// </summary>
+        /// <param name="args">The  object specifically MakeBucketArgs class, args with populated values. From Input</param>
+        /// <param name="method">The Method needed to call with this request object</param>
+        /// <returns>A RestRequest</returns>
+        internal RestRequest CreateRequest(MakeBucketArgs args, Method method)
```
Comment (kannappanr) [inline Minio/MinioClient.cs:155]: why is this `MakeBucketArgs`?

File: Minio/ApiEndpoints/BucketOperations.cs (hunk 29 above)
[the remark this answers is not in the document]
Comment (kannappanr) [reply Minio/ApiEndpoints/BucketOperations.cs:96]: why do we need `updateArgs` ?

File: Minio/DataModel/GenericResponse.cs (hunk 31 above)
Comment (BigUstad) [reply Minio/DataModel/GenericResponse.cs:25]: Bucket name, object name or other info.

File: Minio/ApiEndpoints/BucketOperations.cs (hunk 29 above)
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/ApiEndpoints/BucketOperations.cs]: Removed it.

File: Minio/DataModel/BucketOperationsReponse.cs [hunk 34]
```diff
@@ -0,0 +1,71 @@
...
+            return this;
+        }
+    }
+
+    public class VersioningConfig: BucketResponse
+    {
+        private VersioningConfiguration CurrentVersioningConfiguration { get; set; }
+        public string Status { get; private set;}
+        public bool MfaDelete { get; private set;}
+        public VersioningConfig(HttpStatusCode statusCode, string responseContent)
```
Comment (poornas) [inline Minio/DataModel/BucketOperationsReponse.cs:38]:
```suggestion
        internal VersioningConfig(HttpStatusCode statusCode, string responseContent)
```

File: Minio/DataModel/MinioClientBuilder.cs [hunk 35]
```diff
@@ -0,0 +1,188 @@
...
+            return WithEndpoint(url.AbsoluteUri);
+        }
+
+        public MinioClient WithRegion(string region)
+        {
+            if (String.IsNullOrEmpty(region))
+            {
+                throw new ArgumentException(String.Format("{0} the region value can't be null or empty.", region),"region");
+            }
+            this.RegionInUrl = region;
```
Comment (poornas) [inline Minio/DataModel/MinioClientBuilder.cs:139]:
```suggestion
            this.Region = region;
```

File: Minio/DataModel/MinioClientBuilder.cs [hunk 36]
```diff
@@ -0,0 +1,188 @@
...
+    }
+
+    public partial class MinioClient : IMinioClient
+    {
+        internal bool IsAwsHost { get; private set; }
+        internal bool IsAwsChinaHost { get; private set; }
+        internal bool IsAcceleratedHost { get; private set; }
+        internal bool IsDualStackHost { get; private set; }
+        internal bool UseVirtualStyle { get; private set; }
+        internal string RegionInUrl { get; private set; }
```
Comment (poornas) [inline Minio/DataModel/MinioClientBuilder.cs:42]: RegionInUrl - why are these defined here? Most of the vars defined above are not being used - and can be removed.

File: Minio/MinioClient.cs (hunk 33 above)
Comment (poornas) [reply Minio/MinioClient.cs:155]: if MakeBucket needs to be treated specially, do so within MakeBucketAsync - this generic method is not required.

File: Docs/API.md [hunk 37]
```diff
@@ -245,6 +383,49 @@ catch (MinioException e)
...
+
+Checks if a bucket exists.
+
+
+__Parameters__
+
+
+|Param   | Type	  | Description  |
+|:--- |:--- |:--- |
+| ``args``  | _BucketExistsArgs_  | BucketExistsArgs Arguments Object which has bucket identifier information - bucket name, region.  |
```
Comment (poornas) [inline Docs/API.md:398]: region should be removed.

File: Docs/API.md [hunk 38]
```diff
@@ -301,6 +482,159 @@ catch(MinioException e)
...
+`Task<VersioningConfiguration> GetVersioningAsync(GetVersioningArgs args, CancellationToken cancellationToken = default(CancellationToken))`
+
+Get versioning information for a bucket.
+
+__Parameters__
+
+
+|Param   | Type	  | Description  |
+|:--- |:--- |:--- |
+| ``args``  | _GetVersioningArgs_  | GetVersioningArgs Arguments Object that has the info like bucket name, region. |
```
Comment (poornas) [inline Docs/API.md:552]: | ``args``  | _GetVersioningArgs_  | GetVersioningArgs takes bucket name as argument |

Comment (poornas) [reply Docs/API.md:552]: suggest remaining args description wording also be made concise.

File: Minio/DataModel/BucketOperationsArgs.cs [hunk 39]
```diff
@@ -0,0 +1,85 @@
...
+ */
+
+using Minio.DataModel;
+using RestSharp;
+
+namespace Minio
+{
+    public class BucketExistsArgs: BucketArgs
+    {
+        private string bucketName;
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/DataModel/BucketOperationsArgs.cs:24]: Inserted by VSCode. Sry for the oversight. Removed.

File: Minio/DataModel/BucketOperationsArgs.cs [hunk 40]
```diff
@@ -0,0 +1,85 @@
...
+
+        public BucketExistsArgs(string bucketName)
+                :base(bucketName)
+        {
+        }
+    }
+
+    public class RemoveBucketArgs : BucketArgs
+    {
+        internal string Region;
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/DataModel/BucketOperationsArgs.cs:34]: Removed.

File: Minio/DataModel/BucketOperationsArgs.cs [hunk 41]
```diff
@@ -0,0 +1,85 @@
...
+
+    public class MakeBucketArgs : BucketArgs
+    {
+        public MakeBucketArgs(string bucketName)
+                    : base(bucketName)
+        {
+        }
+
+        internal string Location { get; set; }
+        internal string Region { get; set; }
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/DataModel/BucketOperationsArgs.cs:50]: Also removed. Using Region from MinioClient object.

File: Minio/DataModel/BucketOperationsReponse.cs (hunk 34 above)
Comment (BigUstad) [reply Minio/DataModel/BucketOperationsReponse.cs:38]: Changed. Not calling it from outside anymore.

File: Minio/DataModel/MinioClientBuilder.cs (hunk 36 above)
Comment (BigUstad) [reply Minio/DataModel/MinioClientBuilder.cs:42]: Removed.

File: Docs/API.md (hunk 37 above)
Comment (BigUstad) [reply Docs/API.md:398]: Changes made.

File: Docs/API.md (hunk 38 above)
Comment (BigUstad) [reply Docs/API.md:552]: Concise wording included. Other stuff removed.

File: Minio/ApiEndpoints/BucketOperations.cs [hunk 42]
```diff
@@ -35,6 +35,101 @@ namespace Minio
...
+        {
+            args.Validate();
+            RestRequest request = await this.CreateRequest(Method.DELETE, args.BucketName).ConfigureAwait(false);
+            await this.ExecuteTaskAsync(this.NoErrorHandlers, request, cancellationToken);
+        }
+
+        /// <summary>
+        /// Create a bucket with the given name.
+        /// </summary>
+        /// <param name="args">MakeBucketArgs Arguments Object that has bucket info like name, region, location. etc</param>
```
Comment (poornas) [inline Minio/ApiEndpoints/BucketOperations.cs:89]:
```suggestion
        /// <param name="args">MakeBucketArgs Arguments Object that has bucket info like name, location. etc</param>
```

File: Minio/ApiEndpoints/BucketOperations.cs [hunk 43]
```diff
@@ -35,6 +35,101 @@ namespace Minio
...
+                RequestFormat = DataFormat.Xml
+            };
+            if (string.IsNullOrEmpty(args.Location))
+            {
+                args.Location = this.Region;
+            }
+            await this.ExecuteTaskAsync(this.NoErrorHandlers, args.BuildRequest(request), cancellationToken);
+        }
+
+        /// <param name="args">GetVersioningArgs Arguments Object that has the info like bucket name, region. </param>
```
Comment (poornas) [inline Minio/ApiEndpoints/BucketOperations.cs:108]:
```suggestion
        /// <param name="args">GetVersioningArgs takes bucket as argument. </param>
```

File: Minio/ApiEndpoints/BucketOperations.cs [hunk 44]
```diff
@@ -35,6 +35,101 @@ namespace Minio
 {
     public partial class MinioClient : IBucketOperations
     {
+        /// <summary>
+        /// Check if a private bucket with the given name exists.
+        /// </summary>
+        /// <param name="args">BucketExistsArgs Arguments Object which has bucket identifier information - bucket name, region</param>
+        /// <param name="cancellationToken">Optional cancellation token to cancel the operation</param>
+        /// <returns> Task </returns>
+        public async Task<bool> BucketExistsAsync(BucketExistsArgs args, CancellationToken cancellationToken = default(CancellationToken))
```
Comment (poornas) [inline Minio/ApiEndpoints/BucketOperations.cs:44]: the signatures of the API operations using args style needs to be defined in IBucketOperations.cs. The deprecated methods such as MakeBucketAsync, RemoveBucketAsync and BucketExistsAsync should internally just call the method using args style so that functional tests run it. The previous call signature needs to be marked as deprecated for these 3 API calls.

Comment (BigUstad) [reply Minio/ApiEndpoints/BucketOperations.cs:44]: Marked all the other calls Obsolete with a suitable comment, same as MinioClient constructor.
